File: backend/app/models/device.py
import enum
from sqlalchemy import String, Boolean, Float, ForeignKey, Enum
from sqlalchemy.orm import Mapped, mapped_column, relationship
from .base import Base


class DeviceType(str, enum.Enum):
    LIGHT = "light"
    AC = "ac"
    OUTLET = "outlet"
    CAMERA = "camera"
    TEMP_SENSOR = "temp_sensor"
    MOTION_SENSOR = "motion_sensor"


class Device(Base):
    __tablename__ = "devices"

    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column(String(100), nullable=False)
    type: Mapped[DeviceType] = mapped_column(Enum(DeviceType), nullable=False)
    is_on: Mapped[bool] = mapped_column(Boolean, default=False)
    # numeric value for sensors (temperature, brightness, etc.)
    value: Mapped[float | None] = mapped_column(Float, nullable=True)
    room_id: Mapped[int] = mapped_column(ForeignKey("rooms.id"), nullable=False)

    room: Mapped["Room"] = relationship(back_populates="devices")
    action_logs: Mapped[list["ActionLog"]] = relationship(back_populates="device")

File: mobile/lib/models/device.dart
enum DeviceType {
  light,
  ac,
  outlet,
  camera,
  temp_sensor,
  motion_sensor;

  static DeviceType fromString(String s) =>
      DeviceType.values.firstWhere((e) => e.name == s, orElse: () => DeviceType.light);

  String get label => switch (this) {
        DeviceType.light => 'Лампа',
        DeviceType.ac => 'Кондиционер',
        DeviceType.outlet => 'Розетка',
        DeviceType.camera => 'Камера',
        DeviceType.temp_sensor => 'Температура',
        DeviceType.motion_sensor => 'Датчик движения',
      };
}

class Device {
  final int id;
  final String name;
  final DeviceType type;
  final bool isOn;
  final double? value;
  final int roomId;

  const Device({
    required this.id,
    required this.name,
    required this.type,
    required this.isOn,
    required this.value,
    required this.roomId,
  });

  factory Device.fromJson(Map<String, dynamic> json) => Device(
        id: json['id'] as int,
        name: json['name'] as String,
        type: DeviceType.fromString(json['type'] as String),
        isOn: json['is_on'] as bool,
        value: (json['value'] as num?)?.toDouble(),
        roomId: json['room_id'] as int,
      );

  Device copyWith({bool? isOn, double? value}) => Device(
        id: id,
        name: name,
        type: type,
        isOn: isOn ?? this.isOn,
        value: value ?? this.value,
        roomId: roomId,
      );
}
